src/V3EmitMk.cpp

```diff
@@ -31,11 +31,557 @@ class EmitMk final {
 public:
     // METHODS
 
+    struct FileOrConcatenatedFilesList final {
+        std::string fileName{};
+        std::vector<std::string> concatenatedFileNames{};
+
+        // Concatenating file score for debugTestConcatenation
+        std::int64_t dbgScore = 0;
+
+        bool isConcatenatingFile() const { return !concatenatedFileNames.empty(); }
+    };
+
+    struct FileNameWithScore final {
+        std::string name;
+        std::int64_t score;
+    };
+
+    // Data of a single work unit used in `singleConcatenatedFilesList()`.
+    struct WorkList final {
+        std::int64_t totalScore{0};
+        std::vector<FileNameWithScore> files{};
+        // Number of buckets assigned for this list. Used only in concatenable lists.
+        int bucketsNum{0};
+        // Indicated whether files on this list can be concatenated.
+        bool isConcatenable{true};
+        // Work list ID for debugging.
+        int dbgId;
+
+        WorkList() = delete;
+        WorkList(int id)
+            : dbgId{id} {}
+    };
+
+    // Debug logging helper: Returns properly indented "* " string.
+    // Used as a prefix of log lines representing list entry.
+    static std::string logMsgPrefixListEntry(unsigned level, unsigned ancestorsWithStatusNum = 0) {
+        UASSERT(level >= ancestorsWithStatusNum,
+                "Level must include number of ancestors with status");
+        const int spacesNum = ancestorsWithStatusNum * 6 + (level - ancestorsWithStatusNum) * 2;
+        const std::string indentSpaces = std::string(spacesNum, ' ');
+        return indentSpaces + "* ";
+    };
+
+    // Debug logging helper: Returns properly indented "* [+] " if status is true,
+    // otherwise "- [ ]". Used as a prefix of log lines representing list entry.
+    static std::string logMsgPrefixListEntryWithStatus(bool status, int level,
+                                                       int ancestorsWithStatusNum = 0) {
+        return logMsgPrefixListEntry(level, ancestorsWithStatusNum) + (status ? "[+] " : "[ ] ");
+    };
+
+    // Debug logging: prints scores histogram
+    static void debugLogScoreHistogram(const std::vector<std::int64_t>& sortedScores) {
+        constexpr int LOG_LEVEL = 6;
+
+        if (debug() < LOG_LEVEL) return;
+
+        constexpr int MAX_BAR_LENGTH = 80;
+        constexpr int MAX_INTERVALS_NUM = 60;
+
+        const auto topScore = sortedScores.back();
+        UINFO(LOG_LEVEL, "Top score: " << topScore << endl);
```

Comment:
1. Where reasonable, use the existing V3Stats so the relevant summary information gets into the __stats.txt file.  At least some information should land there, think about what's best.
2. Otherwise, dump to a new output file similar to .txt debug dumps used other places.